File: Mp3GainMac/FileProgressViewItem.m
//
//  FileProgressViewItem.m
//  MP3GainExpress
//

#import "FileProgressViewItem.h"

@interface FileProgressViewItem ()

@end

@implementation FileProgressViewItem

-(void)setRepresentedObject:(id)representedObject{
    [super setRepresentedObject:representedObject];
    self.IsStarted = NO;
    self.pbStatus.minValue = 0.0;
    self.pbStatus.maxValue = 100.0;
    self.pbStatus.usesThreadedAnimation = YES;
    
    if(NSAppKitVersionNumber < NSAppKitVersionNumber10_10){
        //Work-around for the progress bar not being visible on OS X 10.7
        //I think this is caused by using newer dev tools and couldn't find a fix.
        //So just use a different control instead because this one doesn't disappear.
        [self.pbStatus setHidden:YES];
        [self.pbStatusOld setHidden:NO];
    }
    
    if([representedObject isKindOfClass:[Mp3GainTask class]]){
        self.itemTask = representedObject;
        if(self.itemTask.Files.count > 1){
            [self.pbStatus setIndeterminate:YES];
        }
        self.itemTask.onStatusUpdate = ^(double percentComplete) {
            dispatch_async(dispatch_get_main_queue(), ^{
                if(self.pbStatus){
                    if(!self.IsStarted){
                        if(self.itemTask.Files.count > 1){
                            [self.pbStatus setIndeterminate:NO];
                        }
                        if(NSAppKitVersionNumber >= NSAppKitVersionNumber10_10){
                            [self.pbStatus startAnimation:self];
                        }
                        self.IsStarted = YES;
                    }
                    if(percentComplete == 100.0 && self.itemTask.Files.count > 1){
                        //Go back to indeterminate at the end of Album gain because it takes a few seconds to apply
                        //changes to all files after the scanning has finished.
                        [self.pbStatus setIndeterminate:YES];
                    }
                    else{
                        if(NSAppKitVersionNumber < NSAppKitVersionNumber10_10){
                            [self.pbStatusOld setDoubleValue:percentComplete];
                        }
                        else{
                            self.pbStatus.doubleValue = (double)percentComplete;
                        }
                    }
                }
            });
        };
        
        [_lblFilename setStringValue:[self.itemTask getDescription]];
    }
}

@end

File: Mp3GainMac/FileProgressViewItem.swift
import AppKit

final class FileProgressViewItem: NSCollectionViewItem {
    @IBOutlet weak var lblFilename: NSTextField!
    @IBOutlet weak var pbStatus: NSProgressIndicator!
    var itemTask: MP3GainTask?
    @objc dynamic var isStarted = false

    override var representedObject: Any? {
        didSet {
            if representedObject as? MP3GainTask != itemTask && itemTask != nil {
                itemTask!.onStatusUpdate = nil //We're reusing this row, unhook it from the previous events
            }
            guard isViewLoaded else { return }
            isStarted = false
            pbStatus.minValue = 0.0
            pbStatus.maxValue = 100.0
            pbStatus.usesThreadedAnimation = true

            guard let task = representedObject as? MP3GainTask else {
                itemTask = nil
                return
            }

            itemTask = task
            pbStatus.doubleValue = itemTask?.statusValue ?? 0.0
            if task.files.count > 1 {
                pbStatus.isIndeterminate = true
            }

            task.onStatusUpdate = { [weak self] percentComplete in
                DispatchQueue.main.async {
                    guard let self, self.pbStatus != nil else { return }
                    if !self.isStarted {
                        if task.files.count > 1 {
                            self.pbStatus.isIndeterminate = false
                        }
                        self.pbStatus.startAnimation(self)
                        self.isStarted = true
                    }

                    if percentComplete == 100.0 && task.files.count > 1 {
                        // Go back to indeterminate at the end of Album gain because it takes a few seconds to apply
                        // changes to all files after the scanning has finished.
                        self.pbStatus.isIndeterminate = true
                    } else {
                        self.pbStatus.doubleValue = percentComplete
                    }
                }
            }

            lblFilename.stringValue = task.taskDescription
        }
    }
}
